File: docs/breez-sdk/snippets/swift/BreezSdkSnippets/Sources/ParsingInputs.swift
import BreezSdkSpark
import Foundation

func parseInput(sdk: BreezSdk) async throws {
    // ANCHOR: parse-inputs
    let input = "an input to be parsed..."

    do {
        let inputType = try await sdk.parse(input: input)
        switch inputType {
        case .bitcoinAddress(v1: let details):
            print("Input is Bitcoin address \(details.address)")

        case .bolt11Invoice(v1: let details):
            let amount = details.amountMsat.map { String($0) } ?? "unknown"
            print("Input is BOLT11 invoice for \(amount) msats")

        case .lnurlPay(v1: let details):
            print(
                "Input is LNURL-Pay/Lightning address accepting min/max \(details.minSendable)/\(details.maxSendable) msats)"
            )
        case .lnurlWithdraw(v1: let details):
            print(
                "Input is LNURL-Withdraw for min/max \(details.minWithdrawable)/\(details.maxWithdrawable) msats"
            )

        case .sparkAddress(v1: let details):
            print("Input is Spark address \(details.address)")

        case .sparkInvoice(v1: let invoice):
            print("Input is Spark invoice:")
            if let tokenIdentifier = invoice.tokenIdentifier {
                print("  Amount: \(invoice.amount) base units of token with id \(tokenIdentifier)")
            } else {
                print("  Amount: \(invoice.amount) sats")
            }

            if let description = invoice.description {
                print("  Description: \(description)")
            }

            if let expiresAt = invoice.expiresAt {
                print("  Expires at: \(Date(timeIntervalSince1970: TimeInterval(expiresAt)))")
            }

            if let senderPublicKey = invoice.senderPublicKey {
                print("  Sender public key: \(senderPublicKey)")
            }

        default:
            break  // Other input types are available
        }
    } catch {
        print("Failed to parse input: \(error)")
    }
    // ANCHOR_END: parse-inputs
}

func setExternalInputParsers() async {
    // ANCHOR: set-external-input-parsers
    // Create the default config
    var config = defaultConfig(network: Network.mainnet)
    config.apiKey = "<breez api key>"

    // Configure external parsers
    config.externalInputParsers = [
        ExternalInputParser(
            providerId: "provider_a",
            inputRegex: "^provider_a",
            parserUrl: "https://parser-domain.com/parser?input=<input>"
        ),
        ExternalInputParser(
            providerId: "provider_b",
            inputRegex: "^provider_b",
            parserUrl: "https://parser-domain.com/parser?input=<input>"
        ),
    ]
    // ANCHOR_END: set-external-input-parsers
}

File: docs/breez-sdk/snippets/react-native/parsing_inputs.ts
import {
  defaultConfig,
  Network,
  type BreezSdk,
  InputType_Tags
} from '@breeztech/breez-sdk-spark-react-native'

const parseInputs = async (sdk: BreezSdk) => {
  // ANCHOR: parse-inputs
  const inputStr = 'an input to be parsed...'

  const input = await sdk.parse(inputStr)

  if (input.tag === InputType_Tags.BitcoinAddress) {
    console.log(`Input is Bitcoin address ${input.inner[0].address}`)
  } else if (input.tag === InputType_Tags.Bolt11Invoice) {
    console.log(
      `Input is BOLT11 invoice for ${
        input.inner[0].amountMsat != null ? input.inner[0].amountMsat.toString() : 'unknown'
      } msats`
    )
  } else if (input.tag === InputType_Tags.LnurlPay) {
    console.log(
      'Input is LNURL-Pay/Lightning address accepting min/max ' +
        `${input.inner[0].minSendable}/${input.inner[0].maxSendable} msats`
    )
  } else if (input.tag === InputType_Tags.LnurlWithdraw) {
    console.log(
      'Input is LNURL-Withdraw for min/max ' +
        `${input.inner[0].minWithdrawable}/${input.inner[0].maxWithdrawable} msats`
    )
  } else if (input.tag === InputType_Tags.SparkAddress) {
    console.log(`Input is Spark address ${input.inner[0].address}`)
  } else if (input.tag === InputType_Tags.SparkInvoice) {
    const invoice = input.inner[0]
    console.log('Input is Spark invoice:')
    if (invoice.tokenIdentifier != null) {
      console.log(
        `  Amount: ${invoice.amount} base units of token with id ${invoice.tokenIdentifier}`
      )
    } else {
      console.log(`  Amount: ${invoice.amount} sats`)
    }

    if (invoice.description != null) {
      console.log(`  Description: ${invoice.description}`)
    }

    if (invoice.expiryTime != null) {
      console.log(`  Expiry time: ${new Date(Number(invoice.expiryTime) * 1000).toISOString()}`)
    }

    if (invoice.senderPublicKey != null) {
      console.log(`  Sender public key: ${invoice.senderPublicKey}`)
    }
  } else {
    // Other input types are available
  }
  // ANCHOR_END: parse-inputs
}

const exampleSetExternalInputParsers = async () => {
  // ANCHOR: set-external-input-parsers
  // Create the default config
  const config = defaultConfig(Network.Mainnet)
  config.apiKey = '<breez api key>'

  // Configure external parsers
  config.externalInputParsers = [
    {
      providerId: 'provider_a',
      inputRegex: '^provider_a',
      parserUrl: 'https://parser-domain.com/parser?input=<input>'
    },
    {
      providerId: 'provider_b',
      inputRegex: '^provider_b',
      parserUrl: 'https://parser-domain.com/parser?input=<input>'
    }
  ]
  // ANCHOR_END: set-external-input-parsers
}
